workflow/controller/controller.go

```diff
@@ -392,38 +395,34 @@ func (wfc *WorkflowController) initManagers(ctx context.Context) error {
 	return nil
 }
 
-func (wfc *WorkflowController) runConfigMapWatcher(stopCh <-chan struct{}) {
+func (wfc *WorkflowController) newConfigMapCtrlInformer() cache.SharedIndexInformer {
 	defer runtimeutil.HandleCrash(runtimeutil.PanicHandlers...)
 
 	ctx := context.Background()
-	retryWatcher, err := apiwatch.NewRetryWatcher("1", &cache.ListWatch{
-		WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
-			return wfc.kubeclientset.CoreV1().ConfigMaps(wfc.managedNamespace).Watch(ctx, metav1.ListOptions{})
-		},
+
+	informer := v1.NewConfigMapInformer(wfc.kubeclientset, wfc.GetManagedNamespace(), 20*time.Minute, cache.Indexers{
+		cache.NamespaceIndex: cache.MetaNamespaceIndexFunc,
 	})
-	if err != nil {
-		panic(err)
-	}
-	defer retryWatcher.Stop()
 
-	for {
-		select {
-		case event := <-retryWatcher.ResultChan():
-			cm, ok := event.Object.(*apiv1.ConfigMap)
-			if !ok {
-				log.Errorf("invalid config map object received in config watcher. Ignored processing")
-				continue
+	informer.AddEventHandler(cache.FilteringResourceEventHandler{
+		FilterFunc: func(obj interface{}) bool {
+			cm, err := meta.Accessor(obj)
+			if err != nil {
+				return false
 			}
-			log.Debugf("received config map %s/%s update", cm.Namespace, cm.Name)
-			if cm.GetName() == wfc.configController.GetName() && wfc.namespace == cm.GetNamespace() {
+			return cm.GetName() == wfc.configController.GetName()
+		},
+		Handler: cache.ResourceEventHandlerFuncs{
+			UpdateFunc: func(_, obj interface{}) {
+				cm := obj.(*apiv1.ConfigMap)
 				log.Infof("Received Workflow Controller config map %s/%s update", cm.Namespace, cm.Name)
 				wfc.UpdateConfig(ctx)
-			}
-			wfc.notifySemaphoreConfigUpdate(cm)
-		case <-stopCh:
-			return
-		}
-	}
+				wfc.notifySemaphoreConfigUpdate(cm)
```

Comment:
@agilgur5, I'm so sorry, but I'm not sure if I understood well your comment. Is it only about keeping `namespace` instead `managedNamespace`?

Reply:
Not entirely.

`notifySemaphoreConfigUpdate` must run on the `managedNamespace`, yes, as all semaphore ConfigMaps are in the managed namespace.

Further, your `FilterFunc` limits the events to only the Controller's own ConfigMap, while `notifySemaphoreConfigUpdate` needs to run on _all_ ConfigMaps in the managed namespace.

The previous code before your modifications was kind of strangely coupled together (and had a bug for managed namespaces) as those are unrelated features, so I imagine you may have missed that one-line nuance

Reply:
@juranir there's some complexity here since [this](https://github.com/argoproj/argo-workflows/pull/11799) PR is also out at the same time. Both of you seem to have covered part of the equation, but then @agilgur5 has also given a lot of thought on how to best optimize and consolidate code [here](https://github.com/argoproj/argo-workflows/pull/11799#issuecomment-1742309369). It sounds like @agilgur5 is willing to take both of these PRs forward unless you would like to go through his comments in that issue and try to take it on yourself? (@agilgur5 please let me know if I'm mistaken)

Reply:
yep that's correct. @juranir seemed interesting, so I'm very willing to assist him as well 🙂

Reply:
Thank you very much for all your help, @agilgur5  is very nice.

If I understand correctly, the idea is to refactor the current `newConfigMapInformer` function to handle both operations (plugins and the cm controller itself), so I think this [commit](https://github.com/argoproj/argo-workflows/pull/11855/commits/1a6110a7fae92ff3a6877f114f4c8564f4079164) might be on the way (I hope in the right one).

I also added a log as commented by @juliev0 

PS. One test failed, but I'm not sure about the reason or if it's related.